# scCardiacOrganoid
Time-course analysis of iPSC-derived cardiac organoid differentiation

![Project overview](TODO.png)

### Collaboration between the Ma (Syracuse) and Cosgrove (Cornell) labs


### Subdirectory descriptions:
- `scripts`: All the other code we used! Mostly contains R scripts and utility functions used in our spatial analyses
- `references`: info & scripts on reference genomes & annotations used in this study
- `resources`: assorted metadata, gene lists, and other information that we used to analyze our data
*see README files in each subdirectory for more details*

### Data generated for this paper:
- scRNAseq datasets: [GSE###](TODO)

### Cell-cell interaction tools
- [`omnipath`](https://omnipathdb.org/)
- [`cellinker`](http://www.rna-society.org/cellinker/Download.html)

### Software used

scanpy                    1.9.1 
scrublet                  0.2.3
soupx
phate                     1.0.9 

### Contact:
- Zhen Ma (zma112@syr.edu)
- Benjamin D. Cosgrove (bdc68@cornell.edu)
- David W. McKellar (dwm269@cornell.edu)
- Plansky Hoang (phoang@syr.edu)
